### PR TITLE
Ear 427 comments notification

### PR DESCRIPTION
**Pull request by farres1**
> ### BEFORE MAKING YOUR PR
>
> Please ensure:
>
> - There are no linting errors, all tests must pass
> - PR is named after JIRA ticket number e.g. EAR-###
> - **Accesibility** checks are completed:
>   - Elements have discernible and consistent focus states
>   - Elements can be navigated to and used by just a keyboard
>   - Elements and text can be read out by a screen reader, and the descriptions are understandable
> - Your feature / bug fix works across **GCP** and **AWS** (where appropriate)
>   - Are modifications to eq-publisher-v3 required?
>   - Are modifications to the Firestore data source required?

---

# What is the context of this PR?

https://collaborate2.ons.gov.uk/jira/browse/EAR-427

Adds a notification icon when a new, unread comment has been sent by another user in Author.

The notification icon is displayed next to the page on the left navigation menu, and on the tab for the page that has the new comment (the preview tab).

# How to review
To test comment notifications, two users must be signed in at the same time. This can be done by creating a new user, opening Author in an incognito window in Chrome (command+shift+n), and signing in as the new user in this window.

When a new comment is created, a notification icon will be displayed on the other account, next to the page and on the page's preview tab - this icon will not be displayed on the account of the comment's author, only for the other account.

### All page types
- [ ] Introduction
- [ ] Section
- [ ] Question page
- [ ] Confirmation page
- [ ] Calculated summary page
- [ ] Submission page

## Check:

**Unread comments - comments panel**
- [ ] Unread comments are highlighted in green in the comments panel upon opening the preview page - the design is different to the prototype's original design, this new design has been agreed with Joe

**Comment notifications**
- [ ] A comment notification is displayed next to all page types in the left navigation menu when a new comment is created by a different user - **check for all page types listed under the "All page types" heading**
- [ ] A comment notification is not displayed for the author of the comment in the left navigation menu - **check for all page types listed under the "All page types" heading**
- [ ] A comment notification is displayed next to the preview tab for all page types when a new comment is created by a different user - **check for all page types listed under the "All page types" heading**
- [ ] A comment notification is not displayed for the author of the comment next to the preview tab - **check for all page types listed under the "All page types" heading**
- [ ] A comment notification is displayed on the left navigation menu and preview tab if an existing comment is edited - **check for all page types listed under the "All page types" heading**

**Dynamic comments**
- [ ] Comment notifications are displayed dynamically and without refresh in the left navigation menu
- [ ] Comment notifications are displayed dynamically and without refresh next to the preview tab
    - [ ] Test on design page, preview page, and logic page for all page types listed under the "All page types" heading
- [ ] When viewing the preview page, new unread comments are highlighted in green in the comments panel dynamically and without refresh
- [ ] When viewing the preview page, if an existing comment or reply is edited, the edited comment or reply is highlighted in green

**Notification display**
- [ ] The comment notification is displayed on the preview tab while viewing the preview page - different to the prototype's original design, this has been agreed with Joe
- [ ] The comment notification is displayed on the left navigation menu item while viewing the preview page - different to the prototype's original design, this has been agreed with Joe

**Validation errors**
- [ ] If a question page has both validation errors and unread comments, the icons are displayed side by side, with the validation error badge on the right

**Replies**
- [ ] If a reply thread has a new comment, the thread is opened by default when navigating onto the preview page
    - The thread will not be opened by default if a new reply is created while viewing the preview page

**Read comments**
- [ ] The notification icon is removed from the left navigation menu item and preview tab after the comment has been read (after navigating from the preview tab to a different tab on the same page)
- [ ] The notification icon is removed from the left navigation menu item and preview tab after the comment has been read (after navigating to a different page)

**Roles**
- [ ] Comment notifications are visible to the questionnaire's creator
- [ ] Comment notifications are visible to the questionnaire's editors
- [ ] Comment notifications are visible to users who do not have editor access to the questionnaire

**Migration**
1 - Create a questionnaire on a different branch
2 - Add comments to pages
3 - Add editor access to another account through the sharing page
4 - Create a comment using the account that has been given editor access
5 - Change to the comments notification branch
6 - The questionnaire creator's account and the questionnaire editor's account should not receive notifications for the added comments

### What to do after everything is green

1. - [ ] Bring this branch up-to-date with Origin/Master
2. - [ ] If there are a lot of commits or some are not helpful to read, squash them down
3. - [ ] Click the **Merge** button on this pull request
4. - [ ] Does this change mean the **Capability examples** questionnaire should be updated?
5. - [ ] Move the Jira ticket for this task into the next stage of the process

### CODE REVIEW

File: eq-author-api/schema/resolvers/base.js [hunk 1]
```diff
@@ -1075,10 +1076,40 @@ const Resolvers = {
...
+            reply.readBy = [];
+          }
+          if (!reply.readBy.includes(userId)) {
+            reply.readBy.push(userId);
+          }
+        });
+      });
+      await saveComments(questionnaireComments);
+
+      return pageComments;
```
Comment (farres1) [inline eq-author-api/schema/resolvers/base.js:1111]: Add tests

File: eq-author-api/schema/resolvers/base.js [hunk 2]
```diff
@@ -1349,6 +1383,10 @@ const Resolvers = {
   QuestionnaireIntroduction: {
     validationErrorInfo: ({ id }, _, ctx) =>
       returnValidationErrors(ctx, id, ({ type }) => type === "introduction"),
+    comments: ({ id }, args, ctx) => ctx.comments[id],
+  },
+  Submission: {
+    comments: ({ id }, args, ctx) => ctx.comments[id],
```
Comment (farres1) [inline eq-author-api/schema/resolvers/base.js:1389]: Add tests

File: eq-author-api/schema/resolvers/base.js [hunk 3]
```diff
@@ -1366,6 +1404,7 @@ const Resolvers = {
         ({ sectionId, folderId, pageId }) =>
           id === sectionId && !pageId && !folderId
       ),
+    comments: ({ id }, args, ctx) => ctx.comments[id],
```
Comment (farres1) [inline eq-author-api/schema/resolvers/base.js:1407]: Add tests

**Review by martyncolmer — APPROVED** [review]

**Review by jasehumphr — APPROVED** [review]
Sweet

**Pull request merged**